### PR TITLE
Allow caching of HFShowerLibrary

### PR DESCRIPTION
**Pull request by Dr15Jones**
#### PR description:

- Modernized code related to HFShowerLibrary
- decreased memory size of HFShowerPhoton by remove unnecessary virtualization
- Added ability to cache entire contents of HFShowerLibrary file and share it safely across instances of modules. (Ability is off by default).

#### PR validation:

- Added a unit test for HFShowerLibrary to show with and without caching give same results. Test also shows that all three existing data files can be read without errors.
- Ran demonstrator which produces pileup files by running the GEN+SIM step for the minimum bias within the mixing module. This generated around 80,000 min bias per job (800 minibias/event with 100 events) using 8 threads without issues.

Comparing timing with and without caching showed a 1-2% speed improvement when using caching.

### CODE REVIEW

**Comment by Dr15Jones** [comment]
please test

File: SimG4CMS/Calo/src/HFFibre.cc [hunk 1]
```diff
@@ -12,86 +12,108 @@
...
   // Attenuation length
-  attL = hcalsimpar_->attenuationLength_;
-  nBinAtt = static_cast<int>(attL.size());
-  std::stringstream ss1;
-  for (int it = 0; it < nBinAtt; it++) {
-    if (it / 10 * 10 == it) {
-      ss1 << "\n";
+  nBinAtt_ = static_cast<int>(attL_.size());
+
+  edm::LogVerbatim("HFShower").log([&](auto& logger) {
```
Comment (Dr15Jones) [inline SimG4CMS/Calo/src/HFFibre.cc:45]: `LogVerbatim::log` will only call the lambda if the message will potentially be sent. This avoids building the `std::stringstream` only to have its contents thrown away.

File: SimG4CMS/Calo/src/HFShowerLibrary.cc [hunk 2]
```diff
@@ -19,58 +19,117 @@
...
+    params.cacheBranches_ = hfShowerLibrary.getUntrackedParameter<bool>("cacheBranches", false);
+    return params;
+  }
+}  // namespace
+
+HFShowerLibrary::HFShowerLibrary(const HcalDDDSimConstants* hcons,
+                                 const HcalSimulationParameters* hps,
+                                 edm::ParameterSet const& hfShower,
+                                 edm::ParameterSet const& hfShowerLibrary)
+    : HFShowerLibrary(paramsFrom(hfShower, hfShowerLibrary, hcons->getPhiTableHF().front()),
```
Comment (Dr15Jones) [inline SimG4CMS/Calo/src/HFShowerLibrary.cc:73]: Having a constructor that takes the two different ParameterSets allowed those values to be passed to two different functions without having to call `getParameterSet` several times.

File: SimG4CMS/Calo/src/HFShowerLibrary.cc [hunk 3]
```diff
@@ -335,223 +397,299 @@ std::vector<HFShowerLibrary::Hit> HFShowerLibrary::fillHits(const G4ThreeVector&
...
 
-bool HFShowerLibrary::rInside(double r) { return (r >= rMin && r <= rMax); }
+bool HFShowerLibrary::rInside(double r) const { return (r >= rMin_ && r <= rMax_); }
+
+HFShowerLibrary::BranchCache::BranchCache(HFShowerLibrary::BranchReader& iReader, size_t maxRecordsToCache) {
+  auto nRecords = iReader.numberOfRecords();
+  if (nRecords > maxRecordsToCache) {
+    nRecords = maxRecordsToCache;
+  }
+  offsets_.reserve(nRecords + 1);
```
Comment (Dr15Jones) [inline SimG4CMS/Calo/src/HFShowerLibrary.cc:415]: `offsets_[0+record]` returns the starting element in `photons_` to read while `offsets_[1+record]` gives one passed the end. Hence why we need 1 more element.

File: SimG4CMS/Calo/src/HFShowerLibrary.cc [hunk 4]
```diff
@@ -335,223 +397,299 @@ std::vector<HFShowerLibrary::Hit> HFShowerLibrary::fillHits(const G4ThreeVector&
...
+  assert(static_cast<size_t>(iRecord + 1) < offsets_.size());
+
+  auto start = offsets_[iRecord - 1];
+  auto end = offsets_[iRecord];
+
+  return HFShowerPhotonCollection(photons_.begin() + start, photons_.begin() + end);
+}
 
-void HFShowerLibrary::getRecord(int type, int record) {
+HFShowerPhotonCollection HFShowerLibrary::BranchReader::getRecordOldForm(TBranch* iBranch, int iEntry) {
```
Comment (Dr15Jones) [inline SimG4CMS/Calo/src/HFShowerLibrary.cc:466]: Moved handling of the different file formats into their own functions.

**Comment by perrotta** [comment]
+1

**Pull request merged**